COPTER: Failsafe Refactor & FS_OPTIONS

Pedals2Paddles opened the pull request:
(_Updated 2019-10-17_) 
As discussed on a dev call, bitmask options for failsafes was highly desired. Reason being, we can add many many options without adding many many additional parameters.  So One new bitmask parameter called `FS_OPTIONS` now has all the user options for battery, radio, and GCS failsafes in copter.  Now we can have as many little options or modifiers for each type of failsafe as we want.  The user has more flexibility to meet mission needs. And it's pretty straightforward to add new ones.

In the process, it turns out all the failsafes could use a refactoring to clean up the ever expanding nested ifs and missing functionality.

**New bitmask options as of today**
- Continue auto mode on radio failsafe (replaces `FS_THR_ENABLED =  2`)
- Continue auto mode mission on GCS failsafe (replaces `FS_GCS_ENABLED =  2`)
- Continue guided mode on radio failsafe
- Continue landing on radio, GCS, or failsafe if already landing
- Continue in pilot controlled modes on GCS failsafe

**The GCS failsafe** seems have suffered some institutional neglect over the years. I updated that so it can have the same settings, options, and behaviors as the radio failsafe.  Also, I found that the GCS failsafe checks were only running in guided mode.  Not even auto mode, nor any other mode.  The GCS failsafe has been heavily reworked, and will now operate in any mode. Just as before, the checks are bypassed if no GCS has ever been connected.

**The radio failsafe** has been severely cleaned up but isn't operating any differently, other than the addition of FS_OPTIONS.

**Parameter conversion** will set the new continue mission option bits for radio and GCS if your `fs_thr_enable` or `fs_gcs_enable` parameters were previously set to continue mission.  Those two will then be set back to 1 (RTL).

**Autotest refactor** now covers all logic and actions for the radio, GCS, and battery failsafes. Previously, the radio and battery failsafe only test one condition and action.  There was no GCS failsafe test.  I've added a GCS failsafe test. And all three now run through all the possible conditions and actions.  The old autotest for radio and battery failsafe timed at 1.5 minutes.  The new autotest for radio, gcs, and battery takes 13 minutes.  That should give you an idea of how much more detail these tests perform now.

**Benefits**
- Radio failsafe _continue in mission_ only worked with RTL.  You couldn't continue in mission with SmartRTL or Land as a failsafe.  Now you can continue in mission with any of the failsafe actions.
- GCS failsafe is now useful. If your use case requires a GCS for safety, this will allow that.  Where before the GCS failsafe only responding in guided mode. Same options as radio failsafe.
- You can also now choose to ignore the GCS failsafe in pilot controlled modes
- You can choose to ignore GCS failsafe in a mission, which was non-functional before.
- You can ignore radio failsafes while in guided mode, such as when your telemetry has more range than RC and are using guided mode.
- If you are in land mode (flight mode, mission command, or landing phase of RTL), you can choose to continue landing if another failsafe kicks in. Before, you'd be stuck having it do a full RTL procedure when it was just about to land to begin with, wasting battery.
- GCS Failsafe will fall back to RTL if an invalid value is entered into `fs_gcs_enable`.  Previously, if the value was invalid, it would do nothing.

** Wiki Updates **
Done in https://github.com/ArduPilot/ardupilot_wiki/pull/2102

** Blog Post **
I have a blog post ready to post on Discuss that details all of this with pretty tables such.  I will put a link to that blog post in the Facebook group as well.

**TO DO:**
- Mission Planner update for the new bitmasks. Does MP do that little checkbox thing automatically or does MO need to create that?

Pedals2Paddles commented:
@OXINARF I think I did this bitmask thing right, looking at how it was done in Q_Options. But I have never in my life done this, so go easy :)

@WickedShell I know you have some battery monitor stuff coming. But taking a quick glance at it, I don't these will impact each other?

@lvale Look, not for Solo! :)

amilcarlucas commented on ArduCopter/Parameters.cpp:
```diff
@@ -1025,6 +1025,27 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
+    // @Param: FS_THR_OPTIONS
+    // @DisplayName: Radio failsafe options bitmask
+    // @Description: Bitmask of additional options for radio failsafe
+    // @Values: 0:Continue when in Auto mode, 1:Continue when in Guided mode, 2:Continue landing in Land mode
```
do not use `@Values:` use `@Bitmask:` instead. The values themselves are correct already, no need to change those

amilcarlucas commented on ArduCopter/Parameters.cpp:
```diff
@@ -1025,6 +1025,27 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
+    // @Param: FS_BATT_OPTIONS
+    // @DisplayName: Battery failsafe options bitmask
+    // @Description: Bitmask of additional options for battery failsafe
+    // @Values: 0:Continue landing in Land mode, 1:Continue flying in pilot controlled modes
```
do not use `@Values:` use `@Bitmask:` instead. The values themselves are correct already, no need to change those

amilcarlucas commented on ArduCopter/Parameters.cpp:
```diff
@@ -1025,6 +1025,27 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
+    // @Param: FS_GCS_OPTIONS
+    // @DisplayName: GCS failsafe options bitmask
+    // @Description: Bitmask of additional options for GCS failsafe
+    // @Values: 0:Continue landing in Land mode, 1:Continue flying in pilot controlled modes
```
do not use `@Values:` use `@Bitmask:` instead. The values themselves are correct already, no need to change those

amilcarlucas commented on ArduCopter/events.cpp:
```diff
@@ -14,10 +14,23 @@ void Copter::failsafe_radio_on_event()
     } else {
-        if (control_mode == AUTO && g.failsafe_throttle == FS_THR_ENABLED_CONTINUE_MISSION) {
-            // continue mission
+        if (control_mode == AUTO && g2.fs_thr_options & ParametersG2::FS_THR_OPT_MISSION_CONTINUE) {
```
Use parentesis for readability around (g2.fs_thr_options & ParametersG2::FS_THR_OPT_MISSION_CONTINUE)

amilcarlucas commented on ArduCopter/events.cpp:
```diff
@@ -14,10 +14,23 @@ void Copter::failsafe_radio_on_event()
+            // continue mission because the FS_THR_OPTION is set to continue mission
+            gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FS - CONTINUING MISSION");
+
+        } else if (control_mode == GUIDED && g2.fs_thr_options & ParametersG2::FS_THR_OPT_GUIDED_CONTINUE) {
```
please add parentesis

amilcarlucas commented on ArduCopter/events.cpp:
```diff
@@ -14,10 +14,23 @@ void Copter::failsafe_radio_on_event()
+            // continue in guided mode because the FS_THR_OPTION is set to continue in land mode
+            gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FS - CONTINUING GUIDED MODE");  
+
+        } else if (control_mode == LAND && g2.fs_thr_options & ParametersG2::FS_THR_OPT_LAND_CONTINUE) {
```
Please add parentesis

amilcarlucas commented on ArduCopter/events.cpp:
```diff
@@ -54,9 +67,19 @@ void Copter::failsafe_battery_event(void)
+                // continue landing because FS_BATT_OPTION is set to continue when in land mode
+                gcs().send_text(MAV_SEVERITY_WARNING, "BATT FS - CONTINUING LAND MODE");
+                
+            } else if (!AP_Notify::flags.autopilot_mode && g2.fs_batt_options & ParametersG2::FS_BATT_OPT_PILOT_CONTROL_CONTINUE) {
```
Should it not be avoided to use the AP_Notify::flags ?

amilcarlucas commented on ArduCopter/events.cpp:
```diff
@@ -113,10 +136,24 @@ void Copter::failsafe_gcs_check()
-            // continue mission
-        } else if (g.failsafe_gcs != FS_GCS_DISABLED) {
-            set_mode_RTL_or_land_with_pause(MODE_REASON_GCS_FAILSAFE);
+        if (!AP_Notify::flags.autopilot_mode && g2.fs_gcs_options & ParametersG2::FS_GCS_OPT_PILOT_FLIGHT_CONTINUE) {
```
parentesis

amilcarlucas replied:
and AP_Notify::flags

amilcarlucas commented on ArduCopter/events.cpp:
```diff
@@ -113,10 +136,24 @@ void Copter::failsafe_gcs_check()
+            // continue because FS_GCS_OPTION is set to continue in pilot controlled modes
+            gcs().send_text(MAV_SEVERITY_WARNING, "GCS FS - CONTINUING PILOT CONTROL");
+            
+        } else if (control_mode == LAND && g.failsafe_battery_enabled == FS_BATT_LAND && failsafe.battery) {
```
parentesis

amilcarlucas commented on ArduCopter/events.cpp:
```diff
@@ -113,10 +136,24 @@ void Copter::failsafe_gcs_check()
+
+        } else if (control_mode == LAND && g2.fs_gcs_options & ParametersG2::FS_GCS_OPT_LAND_CONTINUE) {
+            // continue landing because the FS_GCS_OPTION is set to continue in land mode
+            gcs().send_text(MAV_SEVERITY_WARNING, "GCS FS - CONTINUING LAND MODE");
```
parentesis

Pedals2Paddles commented:
@amilcarlucas Thanks, I will make those fixups to the parameter files this evening.

Pedals2Paddles commented:
@amilcarlucas updates done.

Pedals2Paddles commented:
Should the _continue in guided mode_ option for radio failsafe have an additional event condition to verify there is actually a GCS connected?  Similarly, should the _continue in pilot controlled modes_ option for GCS and battery failsafe conditions check to make sure an RC is connected?

amilcarlucas commented on ArduCopter/events.cpp:
```diff
@@ -54,9 +67,19 @@ void Copter::failsafe_battery_event(void)
+                // continue landing because FS_BATT_OPTION is set to continue when in land mode
+                gcs().send_text(MAV_SEVERITY_WARNING, "BATT FS - CONTINUING LAND MODE");
+                
+            } else if (!AP_Notify::flags.autopilot_mode && (g2.fs_batt_options & ParametersG2::FS_BATT_OPT_PILOT_CONTROL_CONTINUE)) {
```
If I understood this correctly `AP_Notify::flags` should never be used outside of that module.

amilcarlucas commented on ArduCopter/events.cpp:
```diff
@@ -113,10 +136,24 @@ void Copter::failsafe_gcs_check()
-            // continue mission
-        } else if (g.failsafe_gcs != FS_GCS_DISABLED) {
-            set_mode_RTL_or_land_with_pause(MODE_REASON_GCS_FAILSAFE);
+        if (!AP_Notify::flags.autopilot_mode && (g2.fs_gcs_options & ParametersG2::FS_GCS_OPT_PILOT_FLIGHT_CONTINUE)) {
```
If I understood this correctly `AP_Notify::flags` should never be used outside of that module.

Pedals2Paddles commented:
Good point. I'll use the booleans from the flight modes.

amilcarlucas commented:
Rebase ???

amilcarlucas commented:
@rmackay9  can this one be scheduled for Copter 3.6 milestone ?

Pedals2Paddles commented:
Updated and reconfigured since a lot has changed since back in 2017.  Updated the OP to include current status.

Pedals2Paddles commented:
After sleeping on it, I decided it still needed some work. 

- I got rid of the continue in pilot controlled mode options.  There was no check going the other way if you changed out of a pilot mode. So not worth the complication for now.
- I combined all into a single FS_OPTIONS parameter. This should please all those who shutter every time a new parameter is added :)
- I updated GCS failsafes to have the same settings, options, and behaviors as the radio failsafe.

The bitmask options are now:
- Continue mission on radio failsafe
- Continue mission on GCS failsafe
- Continue guided mode on radio failsafe
- Continue landing on radio failsafe
- Continue landing on battery failsafe
- Continue landing on GCS failsafe

Pedals2Paddles commented:
I found that by design, the GCS failsafe checks were only running in guided mode.  Not even auto mode, nor any other mode. This is contrary to documentation and didn't work with the existing failsafe parameters.  _The GCS failsafe checks will now operate in any mode_. Just as before, the checks are bypassed if no GCS has ever been connected.

Pedals2Paddles commented:
All tested in SITL and working as intended!

Pedals2Paddles commented:
I did the parameter conversion.  It's looking for the new `FS_OPTIONS=0`.  Then if `FS_THR_ENABLE=2` and/or `FS_GCS_ENABLE=2`, it adds them to the new `FS_OPTIONS` bitmask then sets them back to regular RTL.

tridge commented on ArduCopter/Parameters.cpp:
```diff
@@ -1505,3 +1516,32 @@ void Copter::convert_tradheli_parameters(void)
+    }
+
+    // Variable to capture the new FS_OPTIONS setting
+    AP_Int32 fs_options_converted;
```
we shouldn't be declaring an AP_xxx parameter on the stack, just use a int32_t

Pedals2Paddles replied:
Done

peterbarker commented on ArduCopter/Parameters.h:
```diff
@@ -602,6 +602,19 @@ class ParametersG2 {
+
+    // Failsafe options bitmask #34
+    AP_Int32 fs_options;
+    enum {
```
Should be `enum class`

Pedals2Paddles replied:
Done. Thx peter

tridge commented on ArduCopter/Parameters.cpp:
```diff
@@ -1505,3 +1516,32 @@ void Copter::convert_tradheli_parameters(void)
+void Copter::convert_fs_options_params(void)
+{
+    // If FS_OPTIONS is not zero, then it has already been configured and we don't change it.
+    if (g2.fs_options != 0) {
```
should be checking if it is configured, the value 0 could be a real value

tridge commented on ArduCopter/Parameters.cpp:
```diff
@@ -1505,3 +1516,32 @@ void Copter::convert_tradheli_parameters(void)
+    }
+
+    // Update the FS_OPTIONS parameter
+    if (fs_options_converted > 0) {
```
I think we should save regardless of value

tridge commented on ArduCopter/events.cpp:
```diff
@@ -7,29 +7,33 @@
             } else if (g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_SMARTRTL_OR_LAND) {
                 set_mode_SmartRTL_or_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
-            } else { // g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_LAND
+            } else if (g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_LAND) {
```
this is a behaviour change

Pedals2Paddles replied:
Fixed to once again handle invalid values going to land.

rmackay9 commented on ArduCopter/Parameters.cpp:
```diff
@@ -173,7 +173,7 @@ const AP_Param::Info Copter::var_info[] = {
     // @DisplayName: Ground Station Failsafe Enable
     // @Description: Controls whether failsafe will be invoked (and what action to take) when connection with Ground station is lost for at least 5 seconds. NB. The GCS Failsafe is only active when RC_OVERRIDE is being used to control the vehicle.
-    // @Values: 0:Disabled,1:Enabled always RTL,2:Enabled Continue with Mission in Auto Mode,3:Enabled always SmartRTL or RTL,4:Enabled always SmartRTL or Land
+    // @Values: 0:Disabled,1:Enabled always RTL,2:Deprecated,3:Enabled always SmartRTL or RTL,4:Enabled always SmartRTL or Land,5:Enabled always land
```
This will change what users see when running 3.6 (and earlier) so we should add a comment like "Deprecated in 4.0".

Pedals2Paddles replied:
Done.

peterbarker commented on ArduCopter/events.cpp:
```diff
@@ -7,29 +7,33 @@
-            if (g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_RTL) {
-                set_mode_RTL_or_land_with_pause(MODE_REASON_RADIO_FAILSAFE);
-            } else if (g.failsafe_throttle == FS_THR_ENABLED_CONTINUE_MISSION) {
+            if (g.failsafe_throttle == FS_THR_ENABLED_ALWAYS_RTL || g.failsafe_throttle == FS_THR_ENABLED_CONTINUE_MISSION) {
```
This should be a switch, I think.  with a default for unknown values.

Pedals2Paddles replied:
Done. Changed both GCS and radio to switches.

Pedals2Paddles commented:
Numerous updates after today's dev call.  Cleaned up a bunch of old if-else pyramids to nice near switches.  Fixed the invalid value default for radio failsafe.  Added invalid value default to GCS failsafe (RTL).  Peter fixed up the enumeration structure to be a class.  Just need to update the parameter conversion with Tridge this week.

Pedals2Paddles commented:
Did a major cleanup of the radio and GCS failsafe event handling.  It was a tangled web of nested ifs with a switch thrown in the middle of them.  I redid all of it so they are consistent, neat, and easy to follow.  No nested ifs.  One set of if-else does the conditional stuff based on options.  And a switch afterwards executes it.  Retested everything successfully in SITL.  If this is good, I'll squash it down as well.

Pedals2Paddles commented:
I did a major refactoring of the copter autotest.  The autotest will now run through all logic and actions of the radio, GCS, and battery failsafes. This will allow it to catch inadvertent changes elsewhere in the code that impact the failsafe's expected behaviors.

amilcarlucas commented:
This is out of scope for this PR, but I would like to ask for it..... how about introducing "companion computer failsafes" ? Right now there is no distinction between companion computer and GCS, but I think there should, it would allow for safer setups.

rmackay9 commented on ArduCopter/Parameters.cpp:
```diff
@@ -237,7 +237,7 @@ const AP_Param::Info Copter::var_info[] = {
     // @Param: FS_THR_ENABLE
     // @DisplayName: Throttle Failsafe Enable
     // @Description: The throttle failsafe allows you to configure a software failsafe activated by a setting on the throttle input channel
-    // @Values: 0:Disabled,1:Enabled always RTL,2:Enabled Continue with Mission in Auto Mode,3:Enabled always Land,4:Enabled always SmartRTL or RTL,5:Enabled always SmartRTL or Land
```
Option 2. should stay for a while because 3.6 (and earlier users) will still be able to use this version..

Pedals2Paddles replied:
Fixed.

peterbarker commented on Tools/autotest/arducopter.py:
```diff
@@ -415,89 +468,440 @@ def fly_loiter_to_alt(self):
+        self.set_parameter('SIM_GPS_DISABLE', 1)
+        self.mavproxy.expect("SmartRTL deactivated: bad position")
+        self.set_parameter('SIM_GPS_DISABLE', 0)
+        self.wait_ekf_happy
```
Needs some prens

Pedals2Paddles replied:
Fixed.

Pedals2Paddles commented:
Parameter conversion done and tested in SITL.

WickedShell commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,233 @@
 }
 
-void Copter::handle_battery_failsafe(const char *type_str, const int8_t action)
+void Copter::handle_battery_failsafe(const char *type_str, int8_t action)
```
Don't discard const here, you don't match the function signature anymore.

Pedals2Paddles replied:
Fixed

Pedals2Paddles commented:
Oct 7 2019 Dev Call:  Some discussion and overview of the concepts.  No new issues.  All appears to be in good shape.  @rmackay9 will review in the next two weeks.

rmackay9 commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,235 @@
+    } else if (flightmode->is_landing() && failsafe_option(FailsafeOption::THR_LAND_CONTINUE)) {
+        // Allow landing to continue when FS_OPTIONS is set to continue landing
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - CONTINUING LANDING");
+        action = FS_THR_ENABLED_ALWAYS_LAND;
```
In general I think we should consolidate on "Radio Failsafe" instead of "RC" or "Throttle" failsafe.  This enum may be pre-existing though so this isn't a blocker.

Pedals2Paddles replied:
I agree.  The terms radio, rc, and throttle are used interchangeably throughout the code and documentation. That is an existing enum. I would certainly be in favor of making it all consistent.  I would actually favor RC over radio since RC is more specific than radio. Radio could be too open ended... which radio, i have like 4 on my copter :).

This would mean changing EVERYTHING.  Even the existing parameter from `FS_THR_ENABLE` to something like `FS_RC_FAILSAFE`.  I can do this, as long as that amount of changing things is desired.

Pedals2Paddles replied:
`FS_THR_...` is used about 100 times throughout the code. So I'm not this would be a good idea to change to something else.

rmackay9 commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,235 @@
+    int16_t action;
+    action = g.failsafe_throttle; //The action set in FS_THR_ENABLE
+
+    // Conditions to deviate from FS_THR_ENABLE selection and sent specigic GCS warning
```
little typo here, "specigic"

Pedals2Paddles replied:
Fixed

rmackay9 commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,235 @@
-    AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_RADIO, LogErrorCode::FAILSAFE_OCCURRED);
-
+    // Execute the specified action
+    switch(action) {
```
nitpick: some missing spaces here and there like between "switch" and the following "("

Pedals2Paddles replied:
Fixed

rmackay9 commented on ArduCopter/Parameters.cpp:
```diff
@@ -932,6 +932,14 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
+    // @Param: FS_OPTIONS
+    // @DisplayName: Failsafe options bitmask
+    // @Description: Bitmask of additional options for battery, radio, & GCS failsafes. 0 (default) disables all options.
+    // @Values: 0:Disabled, 1:Continue mission on radio FS only, 2:Continue mission on GCS FS only, 3:Continue mission in radio and GCS FS, 67:Continue mission in radio and GCS FS + continue in pilot operated modes
```
I think it would be good if the FC_OPTIONS param values and bitmasks items always started with the type of FS affected by the option.  I.e. "Radio failsafe continue in Auto"

Pedals2Paddles replied:
I tried this but I had a hard time getting them to make sense in context. Left them as is for now.

rmackay9 commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,235 @@
+    // Conditions to deviate from FS_THR_ENABLE selection and sent specigic GCS warning
     if (should_disarm_on_failsafe()) {
+        // should immediately disarm when we're on the ground
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - DISARMING");
```
I wonder if we really want the extra send_texts?  MP already has a pretty easy to read display.  In general I prefer we limit the use of send_text but this isn't a blocker if people think it's really necessary.

Pedals2Paddles replied:
The GCS tells you that you're in failsafe, but not why.  Now with these options, IMO it is important for the operator to be aware of what the copter is doing.

rmackay9 commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,235 @@
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - CONTINUING AUTO MODE MISSION");       
+        action = FS_THR_DISABLED;
+
+    } else if ((control_mode == Mode::Number::GUIDED || control_mode == Mode::Number::GUIDED_NOGPS) &&
```
we should use the flightmode->in_guided_mode() instead of checking the modes directly

Pedals2Paddles replied:
Done

rmackay9 commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,235 @@
+
+    // Conditions to deviate from FS_GCS_ENABLE parameter setting
+    if (!motors->armed()) {
+        action = FS_GCS_DISABLED;
```
The battery failsafe does something very similar but it uses the Failsafe_Action enum.  We could consolidate on the battery failsafe method by adding another switch.  what do you think?

Pedals2Paddles replied:
Not sure what you mean?

Pedals2Paddles commented:
@rmackay9 I agree with pretty much everything you said and I will make these adjustments.  I replied to a few comments where I differ or have further discussion.

Pedals2Paddles commented on ArduCopter/Copter.h:
```diff
@@ -629,6 +629,17 @@ class Copter : public AP_HAL::HAL::Callbacks {
     };
 
+    enum class FailsafeOption{
+        THR_MISSION_CONTINUE=(1<<0),        //1
```
Done. I went with RC instead of radio since lots of things could be radios.

Pedals2Paddles commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,235 @@
+    // Conditions to deviate from FS_THR_ENABLE selection and sent specigic GCS warning
     if (should_disarm_on_failsafe()) {
+        // should immediately disarm when we're on the ground
+        gcs().send_text(MAV_SEVERITY_WARNING, "RADIO FAILSAFE - DISARMING");
```
Done

amilcarlucas commented:
@rmackay9 is this planned for merge in Copter 4.0 ?

rmackay9 commented:
@amilcarlucas, yes, hopefully

rmackay9 commented:
I thought about this some more and I think may be we can merge this without the changes we discussed on the dev call.  The change wouldn't make any functional change and I think we can leave those changes to a later date when we try to consolidate failsafes across vehicles.

rmackay9 commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,221 @@
+
+    } else if ((flightmode->in_guided_mode()) &&
+      (failsafe_option(FailsafeOption::RC_CONTINUE_IF_GUIDED)) && (g.failsafe_gcs != FS_GCS_DISABLED)) {
+        // Allow guided mode to continue when FS_THR_OPTION is set to continue in guided mode.  Only if the GCS failsafe is enabled.
```
I guess it's FS_OPTIONS instead of FS_THR_OPTION?

Pedals2Paddles replied:
Fixed.

rmackay9 commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,221 @@
+{
+    return (g2.fs_options & (uint32_t)opt);
+}
+// enum Failsafe_Action {
```
I guess we can remove these comments?

Pedals2Paddles replied:
oops.  removed.

rmackay9 commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,221 @@
+        desired_action = FS_GCS_DISABLED;
+
+    } else if (failsafe_option(FailsafeOption::GCS_CONTINUE_IF_PILOT_CONTROL) && !flightmode->is_autopilot()) {
+        // should continue when in a pilot controlled mode because FS_OPTOINS is set to continue in pilot controlled modes
```
slight typo in FS_OPTIONS (O and I are reversed)

Pedals2Paddles replied:
oops. Removed.

Pedals2Paddles replied:
fixed.

Pedals2Paddles commented:
I'll fix those last few items this evening after work.

Pedals2Paddles commented on ArduCopter/events.cpp:
```diff
@@ -4,152 +4,235 @@
+    AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_RADIO, LogErrorCode::FAILSAFE_OCCURRED);
 
+    int16_t action;
+    action = g.failsafe_throttle; //The action set in FS_THR_ENABLE
```
done

Pedals2Paddles commented:
Done.

amilcarlucas commented:
it needs a `git rebase master` yet again

Pedals2Paddles commented:
Done

Pedals2Paddles commented:
Rebased again due to a bunch of changes that went in a few days ago before this got merged.

Pedals2Paddles commented:
Corresponding Wiki pages updated in https://github.com/ArduPilot/ardupilot_wiki/pull/2102.  The Wiki PR should be merged when this one is merged. I have a blog post ready to go for this as well.

rmackay9 commented:
@Pedals2Paddles, I've made some non-functional changes in this branch: https://github.com/rmackay9/rmackay9-ardupilot/tree/pedals-fs-options.  The key change is the "desired_action" variables start life as a Failsafe_Action instead of a uint8_t.

I'm testing this now and hopefully I haven't broken anything you've done.

rmackay9 commented:
@peterbarker, can you confirm you're happy with the autotest changes?

rmackay9 commented on ArduCopter/Parameters.cpp:
```diff
@@ -939,14 +939,22 @@ const AP_Param::GroupInfo ParametersG2::var_info[] = {
+    // @DisplayName: Failsafe options bitmask
+    // @Description: Bitmask of additional options for battery, radio, & GCS failsafes. 0 (default) disables all options.
+    // @Values: 0:Disabled, 1:Continue in auto mode on RC FS only, 2:Continue auto mode on GCS FS only, 3:Continue in auto mode on RC and/or GCS FS, 19:Continue in auto mode on RC and/or GCS FS + continue in pilot controlled modes
+    // @Bitmask: 0:Continue if in auto mode on RC failsafe, 1:Continue if in auto mode on GCS failsafe, 3:Continue if in guided mode on RC failsafe, 4:Continue if landing on any failsafe, 5:Continue if in pilot controlled modes on RC failsafe
```
For the FS_OPTIONS parameter Bitmask, is "5" actually "Continue if in pilot controlled modes on GCS failsafe" (instead of RC?").  I've added a fix along with some other slight wording changes to my branch

rmackay9 commented:
I've tested the GCS failsafe on a real vehicle and in general it worked very well.  I've found a few minor issues but I've fixed them in a series of follow-up commits in this branch:  https://github.com/rmackay9/rmackay9-ardupilot/commits/pedals-fs-options

@Pedals2Paddles, do you want to pull in the last few commits from the branch above and make sure you're happy with them?  If yes then I'll go ahead and merge it all to master.  Thanks!

As a side note I've created some issues (https://github.com/ArduPilot/ardupilot/issues/12722, https://github.com/ArduPilot/ardupilot/issues/12723) that would improve the GCS failsafe further but we can handle these as future enhancements because they are existing issues.

Pedals2Paddles commented:
So putting the conversion from parameter value to failsafe action enum value at the beginning instead of the end  if I'm seeing this correctly?  I actually like that.  If we decide to standardize the failsafe parameters across vehicles down the road, this will make the associated code changes very simple. Just delete the block of code with the conversions and everything else is already as it needs to be.

I'll do the AP_Notify lights and buzzer stuff as part of this too. I love lights and buzzers. Surprised I didn't actually do that before everything else.

The pre-arm check I agree should be a thing and will likely annoy people.  I think probably a PR for another day since this PR probably has enough behavior change annoyances in it already.

rmackay9 commented:
Merged, thanks!

Pedals2Paddles commented:
Yay! Thanks all for the input and help.
